aws_lambda_script/app/resources.py

```diff
@@ -0,0 +1,619 @@
+                            partial_project["developed"] = project["developed"]
+                        elif section == "source_control":
+                            partial_project["source_control"] = project[
+                                "source_control"
+                            ]
+                        elif section == "architecture":
+                            partial_project["architecture"] = project["architecture"]
+
+                    filtered_projects.append(partial_project)
+                else:
+                    filtered_projects.append(project)
+
+        return filtered_projects, 200
+
+
+@ns.route("/projects")
+class Projects(Resource):
+
+    # Loop through all projects and return the ones that match the user email in the first user item in the user list
+    @ns.doc(responses={200: "Success", 401: "Authorization is required"})
+    @ns.marshal_with(get_project_model(), as_list=True)
+    def get(self):
+        owner_email = get_user_email(parser.parse_args())
+        data = read_data()
+        user_projects = [
+            proj for proj in data["projects"] if any(user["email"] == owner_email for user in proj["user"])
+        ]
+        return user_projects, 200
+
+    # Add a new project to the list of projects - needs certain fields to be present in the JSON payload, the non required will be saved as null if string or emtpy if list
+    @ns.marshal_list_with(get_project_model())
+    @ns.doc(
+        responses={
+            201: "Created project",
+            401: "Authorization is required",
+            406: "Missing JSON data",
+            409: "Project with the same name and owner already exists",
+        }
+    )
+    def post(self):
+        owner_email = get_user_email(parser.parse_args())
+
+        # Check that required fields are present in the JSON payload
+        new_project = ns.payload
+        print(new_project)
+        if (
+            "user" not in new_project
+            or "details" not in new_project
+            or "email" not in new_project["user"][0]
+            or "name" not in new_project["details"][0]
+            or "stage" not in new_project
+        ):
+            abort(406, description="Missing JSON data")
+
+        # Ensure the email is set to owner_email and add 'Editor' role if email matches owner_email
+        for user in new_project["user"]:
+            if "email" not in user or not user["email"]:
+                user["email"] = owner_email
+            if user["email"] == owner_email:
+                if "roles" not in user:
+                    user["roles"] = []
+                if "Editor" not in user["roles"]:
+                    user["roles"].append("Editor")
+        if not any(user["email"] == owner_email for user in new_project["user"]):
+            new_project["user"].append({"email": owner_email, "roles": ["Editor"], "grade": ""})
+
+        data = read_data()
+
+        for proj in data["projects"]:
```

Comment:
There is a lot of looping going on here!

Reply:
Reduced the nested looping.